### PR TITLE
calculus: Enhance expandability of `periodicity`

### LINKED ISSUE #17995: `periodic` function needs more expandability
I am making a custom function which has periodicity.
I added `period()` method to it, which emulate trigonometric function's `period()` method.

When trigonometric function is passed, `periodic` function calls `period()` method of it. I think it would be nice for it to call that method (if exists) for every type of arguments. (Or, maybe a method named `_eval_period()` can be introduced)

### PR DESCRIPTION
**Pull request by mcpl-sympy**
#### References to other Issues or PRs

Fixes #17995

#### Brief description of what is fixed or changed
If the function has `period` method (as trigonometric function does), it will be used to determine its periodicity.

#### Release Notes

* calculus
    * Added support for periodicity method to determine the periodicity of a Function with period attribute

### CODE REVIEW

File: sympy/calculus/util.py [hunk 1]
```diff
@@ -410,6 +415,30 @@ def periodicity(f, symbol, check=False):
...
+
+    >>> class F(Function):
+    ...     nargs = 1
+    ...     def _period(self, general_period, symbol):  # This emulates trigonometric function's method.
+    ...         arg = self.args[0]
+    ...         if not arg.has(symbol):
+    ...             return S.Zero
+    ...         if arg == symbol:
+    ...             return general_period
+    ...         if symbol is arg.free_symbols:
```
Comment (smichr) [inline sympy/calculus/util.py]:
```suggestion
    ...         if symbol in arg.free_symbols:
```

Comment (mcpl-sympy) [reply sympy/calculus/util.py]: Oh, a silly mistake... Thanks!

**Comment by mcpl-sympy** [comment]
Oh, geez. I was trying to delete a branch in my repository from other PR, and found myself deleted the whole repository.
Will it make this PR unmergable? I can fork a new one and make a PR again since this is a minor change.

**Comment by Sc0rpi0n101** [comment]
If you've only deleted your local copy and have bot changed anything in your fork later, your fork will still be fine. 
You can just clone that again and resume the work from where it was. 

Also, I think  a better entry would be:

Added support for `periodicity` method to determine the periodicity of a `Function` with `period` attribute

Will your change affect the use of the method for a user?
If it's just an internal implementation change and does not affect the end-users, there's no need for a release note entry.

**Comment by mcpl-sympy** [comment]
@Sc0rpi0n101 Could be. With this change, end-user can define "undefined but periodic" function. By the way, your suggestion looks good. I edited the release not entry. Thanks!

**Comment by mcpl-sympy** [comment]
@smichr 
Can you review this, please?

**Comment by mcpl-sympy** [comment]
@oscarbenjamin This PR is not getting any more reviews. Could you please review this?

**Pull request closed without merging**